CMake: Add config-file package support, add static library pkg-config

johannesschaeufele opened the pull request:
With these changes, libpng can be used in CMake projects by using the `find_package` mechanism as follows:
```cmake
find_package(spng)
target_link_libraries(<Target> PRIVATE spng::spng)
target_link_libraries(<Target> PRIVATE spng::spng_static)
```

Also adds generation of the equivalent pkg-config file for the static library

Tachi107 commented on CMakeLists.txt:
```diff
@@ -21,43 +21,73 @@ if(${CMAKE_VERSION} VERSION_GREATER_EQUAL "3.12")
-
 if(NOT CMAKE_HOST_WIN32)
-    set(spng_LIBS -lm ${ZLIB_LIBRARIES})
+    set(MATH_LIBRARY "-lm")
```
This should probably be just `m`, as that'll get translated to the appropriate flag by CMake

Tachi107 commented on CMakeLists.txt:
```diff
@@ -21,43 +21,73 @@ if(${CMAKE_VERSION} VERSION_GREATER_EQUAL "3.12")
     add_definitions( -DSPNG_DISABLE_OPT=1 )
 endif()
 
+set(STATIC_SHARED_TARGETS "")
```
Maybe `spng_TARGETS` is a bit more intuitive?

Tachi107 commented on CMakeLists.txt:
```diff
@@ -21,43 +21,73 @@ if(${CMAKE_VERSION} VERSION_GREATER_EQUAL "3.12")
+foreach(STATIC_SHARED_TARGET ${STATIC_SHARED_TARGETS})
+    target_include_directories(${STATIC_SHARED_TARGET} PUBLIC
+	  $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/spng>
+	  $<INSTALL_INTERFACE:include/>
```
Please use the variables from the `GNUInstallDirs` module:

```suggestion
	  $<INSTALL_INTERFACE:${CMAKE_INSTALL_INCLUDEDIR}/>
```

Tachi107 commented on CMakeLists.txt:
```diff
@@ -21,43 +21,73 @@ if(${CMAKE_VERSION} VERSION_GREATER_EQUAL "3.12")
+	  $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/spng>
+	  $<INSTALL_INTERFACE:include/>
+	)
+	target_link_libraries(${STATIC_SHARED_TARGET} PRIVATE ZLIB::ZLIB)
+    target_link_libraries(${STATIC_SHARED_TARGET} PRIVATE ${MATH_LIBRARY})
```
```suggestion
	target_link_libraries(${STATIC_SHARED_TARGET} PRIVATE ZLIB::ZLIB ${MATH_LIBRARY})
```

Tachi107 commented on CMakeLists.txt:
```diff
@@ -21,43 +21,73 @@ if(${CMAKE_VERSION} VERSION_GREATER_EQUAL "3.12")
+
+set(project_config "${CMAKE_CURRENT_BINARY_DIR}/SPNGConfig.cmake")
+set(targets_export_name SPNGTargets)
+configure_file(cmake/Config.cmake.in "${project_config}" @ONLY)
```
This should probably be replaced with a call to the more specialized [`configure_package_config_file()`](https://cmake.org/cmake/help/latest/module/CMakePackageConfigHelpers.html#command:configure_package_config_file) function. Also part of the `CMakePackageConfigHelpers` module, you may want to use `write_basic_package_version_file()` too in order to generate a Version file which handles versioned dependency lookups

Tachi107 commented on CMakeLists.txt:
```diff
@@ -21,43 +21,73 @@ if(${CMAKE_VERSION} VERSION_GREATER_EQUAL "3.12")
+set(targets_export_name SPNGTargets)
+configure_file(cmake/Config.cmake.in "${project_config}" @ONLY)
+
+set(config_install_dir "lib/cmake/spng")
```
```suggestion
set(config_install_dir "${CMAKE_INSTALL_LIBDIR}/cmake/spng")
```

Tachi107 commented on CMakeLists.txt:
```diff
@@ -66,7 +96,9 @@ if(NOT CMAKE_HOST_WIN32 OR CYGWIN OR MINGW)
     set(includedir  ${CMAKE_INSTALL_FULL_INCLUDEDIR})
     set(LIBS        "-lz -lm")
```
Not touched by your patch, but here `"-lz"` can be dropped, as the .pc file already contains `Requires: zlib` (see https://github.com/randy408/libspng/pull/239)

Tachi107 commented on cmake/Config.cmake.in:
```diff
@@ -0,0 +1,7 @@
+# Allow the usage of [PackageName]_ROOT variables for FindPackage
+if(${CMAKE_VERSION} VERSION_GREATER_EQUAL "3.12")
```
```suggestion
if(CMAKE_VERSION VERSION_GREATER_EQUAL "3.12")
```

Tachi107 reviewed (changes requested):
Thanks for your patch! A few suggestions below.

johannesschaeufele commented:
Incorporated suggested changes

Tachi107 commented on CMakeLists.txt:
```diff
@@ -15,58 +15,99 @@ option(SPNG_STATIC "Build static lib" ON)
 
 # Allow the usage of [PackageName]_ROOT variables for FindPackage
-if(${CMAKE_VERSION} VERSION_GREATER_EQUAL "3.12")
+if(CMAKE_VERSION VERSION_GREATER_EQUAL "3.12")
```
You cannot use `VERSION_GREATER_EQUAL`, since it has been introduced in CMake 3.7 and this project targets CMake 3.0. I know you haven't added `VERSION_GREATER_EQUAL` yourself (it was added in commit 82b0c5107d3ca3bf9386496c5c698f025df92858), but it'd be still nice to fix it.

You could either:

- Bump the cmake_minimum_required version to 3.7 - shouldn't cause issues to anyone, since CMake 3.7 is six years old and this script required it already anyway
- Change the condition to `NOT CMAKE_VERSION VERSION_LESS "3.12"`

Tachi107 commented on cmake/Config.cmake.in:
```diff
@@ -0,0 +1,11 @@
+@PACKAGE_INIT@
+
+# Allow the usage of [PackageName]_ROOT variables for FindPackage
+if(CMAKE_VERSION VERSION_GREATER_EQUAL "3.12")
+    cmake_policy(SET CMP0074 NEW)
+endif()
```
As mentioned above, this requires the usage of CMake 3.7, which might not be desirable in a Config file.

Is setting the policy here required anyway? This seems to me something that should be done by the projects consuming the Config file, not by the file itself. I'd just drop this, but I'm not 100% confident so feel free to prove me wrong.

johannesschaeufele commented:
The behavior from #216 is maintained here, wouldn't lean in favor of making breaking changes with regards to CMake version or how dependencies work, unless this is asked for in this PR by @randy408

Style-wise, I'd lean in favor of ```${CMAKE_VERSION}``` instead of ```CMAKE_VERSION``` and two separate calls to ```target_link_libraries```, but would be happy to change this to whichever version is preferred for this project.

Tachi107 commented:
> The behavior from #216 is maintained here, wouldn't lean in favor of making breaking changes with regards to CMake version or how dependencies work

To be clear, I wasn't suggesting changing any behaviour, nor introducing breaking changes;

- Bumping the minimum required CMake version to 3.7 is not a breaking change, because `VERSION_GREATER_EQUAL` is not available in older versions and these older versions already cannot build the project.
- Modifying `CMAKE_VERSION VERSION_GREATER_EQUAL "3.12"` to `NOT CMAKE_VERSION VERSION_LESS "3.12"` is not a change in behaviour, since "not less" means "greater or equal". This is also safer than bumping the minimum required CMake version.
- As for dropping `cmake_policy(SET CMP0074 NEW)`, I meant to drop it from the Config file only, not the main CMakeLists.txt (and since the Config file is new, not adding the `cmake_policy()` call is neither a breaking nor behavioural change). Leaving the `cmake_policy()` in the Config file means not supporting CMake < 3.7, and likely shouldn't be done by a Config file anyway.

johannesschaeufele commented:
Updated minimum CMake version to 3.12 as per Gitter conversation, which makes setting the policy from #216 superfluous

Reverted to using `"-lm"` instead of `"m"`, as this is consistent with other occurrences, e.g. for pkg-config, and more clearly communicates that we are linking a certain library (whereas that might not immediately be clear for `"m"`)

Tachi107 commented:
> Reverted to using `"-lm"` instead of `"m"`, as this is consistent with other occurrences, e.g. for pkg-config, and more clearly communicates that we are linking a certain library (whereas that might not immediately be clear for `"m"`)

I disagree. `target_link_libraries()` is used to link to libraries, not to pass linker flags, and we are linking to the `m` library, not to the `-lm` one. If you want to explicitly pass a linker flag for consistency, then you should use `target_link_options()`. Also, [according to the documentation](https://cmake.org/cmake/help/latest/command/target_link_libraries.html), passing a flag starting with `-l` is doesn't seem a valid documented usage of the command.

johannesschaeufele commented:
> I disagree. `target_link_libraries()` is used to link to libraries, not to pass linker flags, and we are linking to the `m` library, not to the `-lm` one. If you want to explicitly pass a linker flag for consistency, then you should use `target_link_options()`. Also, [according to the documentation](https://cmake.org/cmake/help/latest/command/target_link_libraries.html), passing a flag starting with `-l` is doesn't seem a valid documented usage of the command.

CMake source code with that very pattern for reference: https://github.com/Kitware/CMake/blob/master/Tests/LinkStatic/CMakeLists.txt

Anyway, made suggested changes

Tachi107 commented:
That's odd. Maybe it's just to test that even the undocumented behaviour doesn't change? Or maybe I read the docs incorrectly? To me, it seems that it explicitly says that `-l` flags are not added to the link options, but it doesn't state that `-l` is stripped and the library is added to the list the request ones; it's just undocumented.

Anyway, after re-reading I realized that it may have sounded harsh, but I really didn't mean to sound like that, sorry :)

johannesschaeufele commented:
> Or maybe I read the docs incorrectly? To me, it seems that it explicitly says that `-l` flags are not added to the link options, but it doesn't state that `-l` is stripped and the library is added to the list the request ones; it's just undocumented.

I would interpret the documentation as stating that `-l` flags fall under the `A plain library name` point (cf. `The library name/flag`) rather than the more generic `A link flag` point, which is concerned with linker flags that are not directly library names (and thus do not go through the conversion mentioned: `The generated link line will ask the linker to search for the library`).

randy408 commented:
Thanks!
